src/libmongoc/src/mongoc/mongoc-openssl.c

```diff
@@ -477,6 +513,142 @@ _get_issuer (X509 *cert, STACK_OF (X509) * chain)
    RETURN (issuer);
 }
 
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+/* OpenSSL 1.1.0+ has conveniences that we polyfill in older OpenSSL versions.
+ */
+
+STACK_OF (X509) * _get_verified_chain (SSL *ssl)
+{
+   return SSL_get0_verified_chain (ssl);
+}
+
+void _free_verified_chain (STACK_OF (X509) * verified_chain)
+{
+   /* _get_verified_chain does not return a copy. Do nothing. */
+   return;
+}
+
+const STACK_OF (X509_EXTENSION) * _get_extensions (const X509 *cert)
+{
+   return X509_get0_extensions (cert);
+}
+
+#else
+/* Polyfill functionality for pre 1.1.0 OpenSSL */
+
+STACK_OF (X509) * _get_verified_chain (SSL *ssl)
+{
+   X509_STORE *store = NULL;
+   X509 *peer = NULL;
+   STACK_OF (X509) *peer_chain = NULL;
+   X509_STORE_CTX *store_ctx = NULL;
+   STACK_OF (X509) *verified_chain = NULL;
+
+   /* Get the certificate the server presented. */
+   peer = SSL_get_peer_certificate (ssl);
+   /* Get the chain of certificates the server presented. This is not a verified
+    * chain. */
+   peer_chain = SSL_get_peer_cert_chain (ssl);
+   store = SSL_CTX_get_cert_store (SSL_get_SSL_CTX (ssl));
+   store_ctx = X509_STORE_CTX_new ();
+   if (!X509_STORE_CTX_init (store_ctx, store, peer, peer_chain)) {
+      MONGOC_ERROR ("failed to initialize X509 store");
+      goto fail;
+   }
+
+   if (X509_verify_cert (store_ctx) <= 0) {
+      MONGOC_ERROR ("failed to obtain verified chain");
+      goto fail;
+   }
+
+   verified_chain = X509_STORE_CTX_get1_chain (store_ctx);
+
+fail:
+   X509_free (peer);
+   X509_STORE_CTX_free (store_ctx);
+   return verified_chain;
+}
+
+/* On OpenSSL < 1.1.0, this chain isn't attached to the SSL session, so we need
+ * it to dispose of itself. */
+void _free_verified_chain (STACK_OF (X509) * verified_chain)
+{
+   if (!verified_chain) {
+      return;
+   }
+   sk_X509_pop_free (verified_chain, X509_free);
+}
+
+const STACK_OF (X509_EXTENSION) * _get_extensions (const X509 *cert)
+{
+   return cert->cert_info->extensions;
+}
+#endif /* OPENSSL_VERSION_NUMBER */
+
+
+#define TLSFEATURE_STATUS_REQUEST 5
+
+/* Parse just enough of a DER encoded data to check if a SEQUENCE of INTEGER
```

Comment:
The server has a full DER parser. We only need to parse this one thing, a SEQUENCE of INTEGER.
I used https://docs.microsoft.com/en-us/windows/win32/seccertenroll/about-sequence as a reference for the expected sequence, and this website: http://lapo.it/asn1js/ to interactively view a parsed cert (here is an example of the [mustStaple cert on that website](http://lapo.it/asn1js/#MIIESDCCAzCgAwIBAgIET-rEjTANBgkqhkiG9w0BAQsFADB0MQswCQYDVQQGEwJVUzERMA8GA1UECAwITmV3IFlvcmsxFjAUBgNVBAcMDU5ldyBZb3JrIENpdHkxEDAOBgNVBAoMB01vbmdvREIxDzANBgNVBAsMBktlcm5lbDEXMBUGA1UEAwwOS2VybmVsIFRlc3QgQ0EwHhcNMjAwMjA2MjAxMzExWhcNNDAwMjA4MjAxMzExWjBiMRAwDgYDVQQKDAdNb25nb0RCMQ8wDQYDVQQLDAZLZXJuZWwxEjAQBgNVBAMMCWxvY2FsaG9zdDELMAkGA1UEBhMCVVMxCzAJBgNVBAgMAk5ZMQ8wDQYDVQQHDAZPQ1NQLTEwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDYXCFeRfhcbEDMXNb5D5MIO4VzejQFyGwWuWpIlYGTh1cZp_RrBASBg0hdy0TkhqrWaLvTUq-SaCGrP8Yg0nYXUXIwMe34QhcaTMLndMpAgCXjjC-KDNLX4210QE2RjKbf-13d_duf8yQIRxYrbumA2q6-qZf8gsSteDYf_JeWQ6IWJ7ttWyT9XpcnWLucfviSkDwQuhaghYHDeaWafXgIViVXfGtl7ivv-qx1JUfRyAADl_vcd_-Z-ZojpOUN7Li38I2O8qKTQ8JuMV80dgBOwH9gf0Ku9R9HXFqbypDdvE22WTKk6cvksR_tK_QTdOCRZWpmKRcnhDkYXc7fr97vAgMBAAGjgfMwgfAwCQYDVR0TBAIwADALBgNVHQ8EBAMCBaAwHQYDVR0lBBYwFAYIKwYBBQUHAwEGCCsGAQUFBwMCMB0GA1UdDgQWBBQ7I-iGMVTuZpVlzy9dyvGNZ8akOzBpBggrBgEFBQcBAQRdMFswLgYIKwYBBQUHMAGGImh0dHA6Ly9sb2NhbGhvc3Q6OTAwMS9wb3dlci9sZXZlbC8wKQYIKwYBBQUHMAGGHWh0dHA6Ly9sb2NhbGhvc3Q6ODEwMC9zdGF0dXMvMBEGCCsGAQUFBwEYBAUwAwIBBTAaBgNVHREEEzARgglsb2NhbGhvc3SHBH8AAAEwDQYJKoZIhvcNAQELBQADggEBAJ6z1nkMhirDDFlDrkJPX23kAh1kqMTbFu2wnnrARzw6JEFz2Tvhx0-ISBjEdsPK7BhAQ_wpGIFriTIatXILqWhUAsZN7vwg2h-xH41rpbDVXIcGQ8jAx1GSZznxoeqH7arqbBWrV2AzhQofr_rSUtkqlZNTHmPIOvMjpVPqDGaC9nnw0gC5Ow8ZqgAqD_TV6Uym14c_vtxUGgOUCo4JVDsa-x4uuN1t2iPHIcfo37BZwa9J8Bw3mJGkxnJonm1OurReQjaNoT3Zgag9B7q0cKFd4YHXa2vmmFszs2LhJI2khB-OB4tp9g581M_VUl6mAyaahHlEzTlA5HiHFNADOug))